### PR TITLE
Use absolute RPATHs while linking due to a binutils bug

### PR DESCRIPTION
**Pull request by nirbheek**
Use `-rpath-link` with the absolute paths to the respective build dirs to work around a binutils bug that causes `$ORIGIN` in `-rpath` to not be used while linking.

Includes a unit test that manually checks the `RPATH` value written out to ensure that it still uses `$ORIGIN`.

See: https://sourceware.org/bugzilla/show_bug.cgi?id=16936

Closes https://github.com/mesonbuild/meson/issues/1897

### CODE REVIEW

**Comment by nirbheek** [comment]
What's remaining is fixing `gnome.generate_gir()` so that it passes `-rpath-link` to `g-ir-scanner`.

```
[11/12] Generating MesonSub-1.0.gir with a custom command.
FAILED: gir/MesonSub-1.0.gir 
/usr/bin/g-ir-scanner ../gir/meson-subsample.c ../gir/meson-subsample.h -pthread -I/usr/include/gobject-introspection-1.0 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include --no-libtool --namespace=MesonSub --nsversion=1.0 --warn-all --output gir/MesonSub-1.0.gir '-I/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/gir' '-I/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/build/gir' -I./mesongir/. -I../mesongir/. --include=GObject-2.0 --include=Meson-1.0 --symbol-prefix=meson_sub_ --identifier-prefix=MesonSub . -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include '-L/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/build/gir' -lgobject-2.0 -lglib-2.0 '--add-include-path=/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/build/mesongir' --library girsubproject
/usr/bin/ld: warning: libgirlib.so, needed by /home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/build/gir/libgirsubproject.so, not found (try using -rpath or -rpath-link)
/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/build/gir/libgirsubproject.so: undefined reference to `meson_sample_get_type'
collect2: error: ld returned 1 exit status
Traceback (most recent call last):
  File "/usr/bin/g-ir-scanner", line 66, in <module>
    sys.exit(scanner_main(sys.argv))
  File "/usr/lib64/gobject-introspection/giscanner/scannermain.py", line 544, in scanner_main
    shlibs = create_binary(transformer, options, args)
  File "/usr/lib64/gobject-introspection/giscanner/scannermain.py", line 419, in create_binary
    gdump_parser.get_error_quark_functions())
  File "/usr/lib64/gobject-introspection/giscanner/dumper.py", line 328, in compile_introspection_binary
    return dc.run()
  File "/usr/lib64/gobject-introspection/giscanner/dumper.py", line 174, in run
    self._link(bin_path, introspection_obj)
  File "/usr/lib64/gobject-introspection/giscanner/dumper.py", line 296, in _link
    raise LinkError(e)
distutils.errors.LinkError: command 'gcc' failed with exit status 1
ninja: build stopped: subcommand failed.
```

The command-line is:

> /usr/bin/g-ir-scanner ../gir/meson-subsample.c ../gir/meson-subsample.h -pthread -I/usr/include/gobject-introspection-1.0 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include --no-libtool --namespace=MesonSub --nsversion=1.0 --warn-all --output gir/MesonSub-1.0.gir '-I/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/gir' '-I/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/build/gir' -I./mesongir/. -I../mesongir/. --include=GObject-2.0 --include=Meson-1.0 --symbol-prefix=meson_sub_ --identifier-prefix=MesonSub . -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include '-L/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/build/gir' -lgobject-2.0 -lglib-2.0 '--add-include-path=/home/nirbheek/projects/repositories/github/meson.git/test cases/frameworks/12 multiple gir/build/mesongir' --library girsubproject

**Comment by nirbheek** [comment]
I think the reason why this doesn't show up on Ubuntu (and probably Debian Sid) is because they use `DT_RUNPATH` instead of `DT_RPATH` and `ld` probably does look at that while looking for libraries at link time.

**Comment by tp-m** [comment]
if generate_gir needs fixing, does gtkdoc need fixing too? (It also builds a binary to introspect GTypes)

**Comment by nirbheek** [comment]
Most likely, yes. I haven't run the gtk-doc test because it's still disabled by default. I'm going to add a unit test that runs it from a path without spaces so that we have some automated coverage for it.

File: mesonbuild/compilers.py [hunk 1]
```diff
@@ -801,10 +779,40 @@ def get_soname_args(self, prefix, shlib_name, suffix, path, soversion, is_shared
...
+                rel_rpaths.append(relative)
+            paths = ':'.join([os.path.join('$ORIGIN', p) for p in rel_rpaths])
+            if len(paths) < len(install_rpath):
+                padding = 'X' * (len(install_rpath) - len(paths))
+                if not paths:
+                    paths = padding
+                else:
+                    paths = paths + ':' + padding
+            args = ['-Wl,-rpath,' + paths]
+            if get_compiler_is_linuxlike(self):
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [F821] undefined name 'self'

File: mesonbuild/compilers.py [hunk 2]
```diff
@@ -801,10 +779,40 @@ def get_soname_args(self, prefix, shlib_name, suffix, path, soversion, is_shared
...
+                # Rpaths to use while linking must be absolute. These are not
+                # written to the binary. Needed only with GNU ld:
+                # https://sourceware.org/bugzilla/show_bug.cgi?id=16936
+                # Not needed on Windows or other platforms that don't use RPATH
+                # https://github.com/mesonbuild/meson/issues/1897
+                lpaths = ':'.join([os.path.join(build_dir, p) for p in rpath_paths])
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [F841] local variable 'lpaths' is assigned to but never used

File: mesonbuild/compilers.py [hunk 3]
```diff
@@ -801,10 +779,40 @@ def get_soname_args(self, prefix, shlib_name, suffix, path, soversion, is_shared
...
+                args += ['-Wl,-rpath-link,' + linkpaths]
```
Comment (ignatenkobrain) [inline mesonbuild/compilers.py]: __[flake8]__
* [F821] undefined name 'linkpaths'

**Comment by nirbheek** [comment]
See https://ci.gstreamer.net/job/GStreamer-master-meson/1423/consoleFull for the gst-build failure.

**Pull request merged**